libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -590,211 +574,92 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
-        // Guard from host visibility when compiling in host only contexts
-        NV_IF_TARGET(
-            NV_IS_DEVICE, (
-                for (_CUDA_VSTD::size_t __offset = __rank * 4; __offset < __total_size; __offset += __stride * 4) {
-                    asm volatile ("cp.async.ca.shared.global [%0], [%1], 4, 4;"
-                        :: "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__destination + __offset))),
-                            "l"(__source + __offset)
-                        : "memory");
-                }
-            )
-        )
-        return async_contract_fulfillment::async;
-    }
-};
-
-template<>
-struct __memcpy_async_impl<8, false> {
-    _LIBCUDACXX_DEVICE static inline async_contract_fulfillment __copy(char * __destination, char const * __source, _CUDA_VSTD::size_t __total_size, _CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __stride) {
-        // Guard from host visibility when compiling in host only contexts
-        NV_IF_TARGET(
-            NV_IS_DEVICE, (
-                for (_CUDA_VSTD::size_t __offset = __rank * 8; __offset < __total_size; __offset += __stride * 8) {
-                    asm volatile ("cp.async.ca.shared.global [%0], [%1], 8, 8;"
-                        :: "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__destination + __offset))),
-                            "l"(__source + __offset)
-                        : "memory");
-                }
-            )
-        )
-        return async_contract_fulfillment::async;
-    }
-};
-
-template<>
-struct __memcpy_async_impl<16, false> {
-    _LIBCUDACXX_DEVICE static inline async_contract_fulfillment __copy(char * __destination, char const * __source, _CUDA_VSTD::size_t __total_size, _CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __stride) {
-        // Guard from host visibility when compiling in host only contexts
-        NV_IF_TARGET(
-            NV_IS_DEVICE, (
-                for (_CUDA_VSTD::size_t __offset = __rank * 16; __offset < __total_size; __offset += __stride * 16) {
-                    asm volatile ("cp.async.cg.shared.global [%0], [%1], 16, 16;"
-                        :: "r"(static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__destination + __offset))),
-                            "l"(__source + __offset)
-                        : "memory");
-                }
-            )
+    template<typename _Arch>
+    _LIBCUDACXX_INLINE_VISIBILITY
+    void operator()(_Arch) {
+        const void * __p = &__b;
+
+        _LIBCUDACXX_HANDLE_POINTER_SPACE(__p,
+            __memcpy_async_sync_hooks<
+                barrier<_Sco, _CompF>, false, _Arch,
+                __space::__shared, __space::__global, __p_space_t::value
+            >::__synchronize(_Arch{}, __b,
+                __arch::__is_cuda_provides_sm<_Arch, 80>::value ? async_contract_fulfillment::async : async_contract_fulfillment::none);
```

Comment:
We should have a helper for that

Reply:
This is literally the only place where this pattern is used, and it's only because of an artifact of how I'm reusing almost-unrelated code to satisfy pipeline's needs.

Reply:
You mean the expansion of the macro? I've tried that, it ends up being messier and harder on the compiler. I've started with a helper, and then with this form within a helper, and eventually ended up going to just straight macro invocations, because they are much easier to work with (especially in the case with many arguments, as in the main feature of this PR).

Reply:
Oh no I meant:
```cpp
__arch::__is_cuda_provides_sm<_Arch, 80>::value ? async_contract_fulfillment::async : async_contract_fulfillment::none
```

Reply:
If that ever shows up elsewhere, then we've done something wrong - this is very much just a deep detail of the interactions between pipeline and barrier.